numba/untyped_passes.py

```diff
@@ -463,3 +468,902 @@ def run_pass(self, state):
+        loop_blocks = {
+            x: func_ir.blocks[x] for x in this_loop_body}
+        new_ir = func_ir.derive(loop_blocks)
+
+        # Work out what is live on entry and exit so as to prevent
+        # replacement (defined vars can escape, used vars live at the header
+        # need to remain as-is so their references are correct, they can
+        # also escape).
+
+        usedefs = compute_use_defs(func_ir.blocks)
+        idx = this_loop.header
+        keep = set()
+        keep |= usedefs.usemap[idx] | usedefs.defmap[idx]
+        keep |= func_ir.variable_lifetime.livemap[idx]
+        dont_replace = [x for x in (keep)]
+
+        # compute the unrolled body
+        unrolled_body = self.inject_loop_body(
+            branches, new_ir, max(func_ir.blocks.keys()),
+            dont_replace, switch_data)
+
+        # 6. Patch in the unrolled body and fix up
+        blks = state.func_ir.blocks
+        orig_lbl = tuple(this_loop_body)
+        data = unrolled_body, this_loop.header
+        # python 2 can't star unpack
+        replace, delete = orig_lbl[0], orig_lbl[1:]
+        unroll, header_block = data
+        unroll_lbl = [x for x in sorted(unroll.blocks.keys())]
+        blks[replace] = unroll.blocks[unroll_lbl[0]]
+        [blks.pop(d) for d in delete]
+        for k in unroll_lbl[1:]:
+            blks[k] = unroll.blocks[k]
+        # stitch up the loop predicate true -> new loop body jump
+        blks[header_block].body[-1].truebr = replace
+
+    def run_pass(self, state):
+        mutated = False
+        func_ir = state.func_ir
+        # first limit the work by squashing the CFG if possible
+        func_ir.blocks = simplify_CFG(func_ir.blocks)
+
+        if self._DEBUG:
+            print("-" * 80 + "PASS ENTRY")
+            func_ir.dump()
+            print("-" * 80)
+
+        # limitations:
+        # 1. No nested unrolls
+        # 2. Opt in via `numba.literal_unroll`
+        # 3. No multiple mix-tuple use
+
+        # TODO: Add the following in so that loop nests are walked in reverse
+        # nesting order and as a result nested transforms become valid.
+        def true_body(loop):
+            # computes blocks reachable from a loop head
+            return loop.body - {loop.header} - loop.exits
+
+        def get_loop_nest(structure):
+            all_children = []
+            for x in structure.values():
+                all_children.extend(x)
+
+            # BFS order
+            def BFS(root):
+                acc = []
+
+                def walk(node):
+                    for x in structure[node]:
+                        acc.append(x)
+                    for x in structure[node]:
+                        walk(x)
+                walk(root)
+                return acc
+
+            nest = {}
+            for x in structure.keys():
+                if x not in all_children:
+                    nest[x] = BFS(x)
+            return nest
+
+        def compute_loop_structure(loops):
```

Comment:
dead code?

Reply:
yeah, so I am not sure whether to keep this, I can put it back so can just delete. Basically these functions work out the loop nests ordering such that this pass could be incrementally applied from bottom of the nest up (nesting is currently banned)!

Reply:
removed, will add back at later date